libcst/metadata/scope_provider.py

```diff
@@ -86,6 +99,84 @@ class BuiltinAssignment(BaseAssignment):
     pass
 
 
+class QualifiedNameSource(Enum):
+    IMPORT = auto()
+    LOCAL = auto()
+
+
+@dataclass(frozen=True)
+class QualifiedName:
+    name: str
+    source: QualifiedNameSource
+
+
+class _QualifiedNameUtil:
+    @staticmethod
+    def get_full_name_for(node: cst.CSTNode) -> Optional[str]:
+        if isinstance(node, cst.Name):
+            return node.value
+        elif isinstance(node, cst.Attribute):
+            return (
+                f"{_QualifiedNameUtil.get_full_name_for(node.value)}.{node.attr.value}"
+            )
+        elif isinstance(node, cst.Call):
+            return _QualifiedNameUtil.get_full_name_for(node.func)
+        elif isinstance(node, cst.Subscript):
+            return _QualifiedNameUtil.get_full_name_for(node.value)
+        elif isinstance(node, cst.SimpleString):
+            # In the case of SimpleString of type hints.
+            return literal_eval(node.value)
+        return None
+
+    @staticmethod
+    def find_qualified_name_for_import_alike(
+        assignment_node: Union[cst.Import, cst.ImportFrom],
+        name_parts: List[str],
+        results: Set[QualifiedName],
+    ) -> None:
+        module = ""
+        if isinstance(assignment_node, cst.ImportFrom):
+            module_attr = assignment_node.module
+            if module_attr:
+                module = _QualifiedNameUtil.get_full_name_for(module_attr)
+        import_names = assignment_node.names
+        if not isinstance(import_names, cst.ImportStar):
+            for name in import_names:
+                real_name = _QualifiedNameUtil.get_full_name_for(name.name)
+                as_name = name.asname or real_name
+                if as_name == name_parts[0]:
+                    if module:
+                        real_name = f"{module}.{real_name}"
+                    if real_name:
+                        results.add(
+                            QualifiedName(
+                                ".".join([real_name, *name_parts[1:]]),
+                                QualifiedNameSource.IMPORT,
+                            )
+                        )
+
+    @staticmethod
+    def find_qualified_name_for_non_import(
+        assignment: Assignment, name_parts: List[str], results: Set[QualifiedName]
+    ) -> None:
+        scope = assignment.scope
+        name_prefixes = []
+        while scope:
+            if isinstance(scope, ClassScope) or isinstance(scope, FunctionScope):
+                name_prefixes.append(scope.name)
+            elif isinstance(scope, GlobalScope):
+                break
+            else:
+                raise Exception(f"Unexpected Scope: {scope}")
+            scope = scope.parent
+
+        results.add(
+            QualifiedName(
+                ".".join([*name_prefixes[::-1], *name_parts]), QualifiedNameSource.LOCAL
```

Comment:
Minor nit: I find `reversed(name_prefixes)` much easier to read than `name_prefixes[::-1]`.

Reply:
Fixed.